src/Storages/PostgreSQL/MaterializedPostgreSQLConsumer.cpp

```diff
@@ -587,10 +589,17 @@ void MaterializedPostgreSQLConsumer::syncTables()
                 executor.execute();
             }
         }
-        catch (...)
+        catch (DB::Exception & e)
         {
-            tryLogCurrentException(__PRETTY_FUNCTION__);
+            if (e.code() == ErrorCodes::TOO_MANY_PARTS)
+            {
+                /// Retry this buffer later.
+                storage_data.buffer.columns = result_rows.mutateColumns();
```

Comment:
Should we retry some other errors, for example, `MEMORY_LIMIT_EXCEEDED`?

Reply:
> it will be anyway retried, just on a different level

So what's the difference between these levels and why should we retry `TOO_MANY_PARTS` on this level?

Reply:
in this loop we sync changes for several tables, they all lie in range (last_commited_lsn, final_commit_lsn], each change corresponds to some number in this range, so we could successfully sync changes for one table with a change with bigger LSN, but failed to sync a change with lower LSN, so we cannot commit only successful changes as the LSNs are mixed between different tables. As LSN is always the same for each change and is equal to version column of ReplacingMergeTree table, then double insertion of the same record will be deduplicated. But here for too many parts I just wanted to avoid this extra work as error is retriable easily and can happen often.

Reply:
ah no, this was a mistake, because of this `if`, other errors will not be retried now 🥺

Reply:
https://github.com/ClickHouse/ClickHouse/pull/52381
yes, it was incorrect to add this check only for `too_many_parts` the buffer will be cleared for other failures and we will just get empty columns on next retry, while actually there was an error. I was mistaken because forgot about these lines: https://github.com/ClickHouse/ClickHouse/pull/49698/files#diff-99ffe321e146d9061ab12679d2ef9316b5349dcb87b372d1be3efc888f4722b8R754-R757
- here we will make a commit for final lsn...